src/databricks/labs/ucx/azure/azure_credentials.py

```diff
@@ -0,0 +1,284 @@
+            )
+            return None
+        except InternalError:
+            logger.info(
+                f"InternalError while reading secret {scope}.{key}. "
+                f"Cannot fetch the service principal client_secret for {application_id}. "
+                f"Will not reuse this client_secret"
+            )
+            print(
+                f"{application_id} is not migrated due to InternalError while fetching the client_secret from Databricks secret."
+                f"You may rerun the migration command later to retry this service principal"
+            )
+            return None
+
+        # decode the bytes string from GetSecretResponse to utf-8 string
+        # TODO: handle different encoding if we have feedback from the customer
+        try:
+            if secret_response.value is None:
+                return None
+            return base64.b64decode(secret_response.value).decode("utf-8")
+        except UnicodeDecodeError:
+            logger.info(
+                f"Secret {scope}.{key} has Base64 bytes that cannot be decoded to utf-8 string . "
+                f"Cannot fetch the service principal client_secret for {application_id}. "
+                f"Will not reuse this client_secret"
+            )
+            return None
+
+    def _fetch_client_secret(self, sp_list: list[StoragePermissionMapping]) -> list[ServicePrincipalMigrationInfo]:
+        # check AzureServicePrincipalInfo from AzureServicePrincipalCrawler, if AzureServicePrincipalInfo
+        # has secret_scope and secret_key not empty, fetch the client_secret and put it to the client_secret field
+        #
+        # The input StoragePermissionMapping may have managed identity mixed in, we will ignore them for now, as
+        # they won't have any client_secret, we will process managed identity in the future.
+
+        # fetch client_secrets of crawled service principal, if any
+        azure_sp_info_with_client_secret: dict[str, str] = {}
+        azure_sp_infos = self._azure_sp_crawler.snapshot()
+
+        for azure_sp_info in azure_sp_infos:
+            if not azure_sp_info.secret_scope:
+                continue
+            if not azure_sp_info.secret_key:
+                continue
+            secret_value = self._read_databricks_secret(
+                azure_sp_info.secret_scope, azure_sp_info.secret_key, azure_sp_info.application_id
+            )
+            if secret_value:
+                azure_sp_info_with_client_secret[azure_sp_info.application_id] = secret_value
+
+        # update the list of ServicePrincipalMigrationInfo if client_secret is found
+        sp_list_with_secret = []
+        for sp in sp_list:
+            if sp.client_id in azure_sp_info_with_client_secret:
+                sp_list_with_secret.append(
+                    ServicePrincipalMigrationInfo(
+                        service_principal=sp, client_secret=azure_sp_info_with_client_secret[sp.client_id]
+                    )
```

Comment:
```suggestion
                    ServicePrincipalMigrationInfo(sp, azure_sp_info_with_client_secret[sp.client_id])
```
don't abuse keyword arguments - they hurt readability

Reply:
fixed